File: trip-kit/src/main/java/com/skedgo/android/tripkit/RegionServiceImpl.java
package com.skedgo.android.tripkit;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.skedgo.android.common.model.Location;
import com.skedgo.android.common.model.ModeInfo;
import com.skedgo.android.common.model.Region;
import com.skedgo.android.common.model.TransportMode;
import com.skedgo.android.tripkit.tsp.Paratransit;
import com.skedgo.android.tripkit.tsp.RegionInfo;
import com.skedgo.android.tripkit.tsp.RegionInfoService;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

import javax.inject.Provider;

import rx.Observable;
import rx.functions.Action0;
import rx.functions.Func1;

final class RegionServiceImpl implements RegionService {
  private final Cache<List<Region>> regionCache;
  private final Cache<Map<String, TransportMode>> modeCache;
  private final RegionsFetcher regionsFetcher;
  private final Provider<RegionInfoService> regionInfoServiceProvider;

  RegionServiceImpl(
      Cache<List<Region>> regionCache,
      Cache<Map<String, TransportMode>> modeCache,
      @NonNull RegionsFetcher regionsFetcher,
      Provider<RegionInfoService> regionInfoServiceProvider) {
    this.regionCache = regionCache;
    this.modeCache = modeCache;
    this.regionsFetcher = regionsFetcher;
    this.regionInfoServiceProvider = regionInfoServiceProvider;
  }

  @Override
  public Observable<List<Region>> getRegionsAsync() {
    return regionCache.getAsync();
  }

  @Override
  public Observable<Map<String, TransportMode>> getTransportModesAsync() {
    return modeCache.getAsync();
  }

  @Override
  public Observable<Region> getRegionByLocationAsync(@Nullable final Location location) {
    if (location == null) {
      return Observable.error(new NullPointerException("Location is null"));
    }

    return getRegionsAsync()
        .flatMap(new Func1<List<Region>, Observable<Region>>() {
          @Override
          public Observable<Region> call(List<Region> regions) {
            return Observable.from(regions);
          }
        })
        .first(new Func1<Region, Boolean>() {
          @Override
          public Boolean call(Region region) {
            return region.containsLocation(location);
          }
        })
        .onErrorResumeNext(new Func1<Throwable, Observable<? extends Region>>() {
          @Override
          public Observable<? extends Region> call(Throwable error) {
            return error instanceof NoSuchElementException
                ? Observable.<Region>error(new OutOfRegionsException(location, "Location lies outside covered area"))
                : Observable.<Region>error(error);
          }
        });
  }

  @Override
  public Observable<Location> getCitiesAsync() {
    return getRegionsAsync()
        .flatMap(new Func1<List<Region>, Observable<Region>>() {
          @Override
          public Observable<Region> call(List<Region> regions) {
            return Observable.from(regions);
          }
        })
        .compose(Utils.getCities());
  }

  @Override
  public Observable<Location> getCitiesByNameAsync(@Nullable String name) {
    return getCitiesAsync().filter(Utils.matchCityName(name));
  }

  @Override
  public Observable<List<TransportMode>> getTransportModesByIdsAsync(@NonNull final List<String> modeIds) {
    return getTransportModesAsync()
        .map(Utils.findModesByIds(modeIds));
  }

  @Override
  public Observable<List<TransportMode>> getTransportModesByRegionAsync(@NonNull Region region) {
    final List<String> modeIds = region.getTransportModeIds();
    return modeIds != null
        ? getTransportModesByIdsAsync(modeIds)
        : Observable.just(Collections.<TransportMode>emptyList());
  }

  @Override
  public Observable<List<TransportMode>> getTransportModesByLocationAsync(@NonNull Location location) {
    return getRegionByLocationAsync(location)
        .flatMap(new Func1<Region, Observable<List<TransportMode>>>() {
          @Override
          public Observable<List<TransportMode>> call(Region region) {
            final List<String> modeIds = region.getTransportModeIds();
            return modeIds != null
                ? getTransportModesByIdsAsync(modeIds)
                : Observable.just(Collections.<TransportMode>emptyList());
          }
        });
  }

  @Override public Observable<Void> refreshAsync() {
    return regionsFetcher.fetchAsync()
        .doOnCompleted(new Action0() {
          @Override public void call() {
            regionCache.invalidate();
            modeCache.invalidate();
          }
        });
  }

  @Override
  public Observable<Paratransit> fetchParatransitByRegionAsync(@NonNull final Region region) {
    return getRegionInfoByRegionAsync(region)
        .map(new Func1<RegionInfo, Paratransit>() {
          @Override public Paratransit call(RegionInfo regionInfo) {
            return regionInfo.paratransit();
          }
        });
  }

  @Override
  public Observable<RegionInfo> getRegionInfoByRegionAsync(@NonNull final Region region) {
    return regionInfoServiceProvider.get()
        .fetchRegionInfoAsync(region.getURLs(), region.getName());
  }

  @Override
  public Observable<List<ModeInfo>> getTransitModesByRegionAsync(@NonNull final Region region) {
    return getRegionInfoByRegionAsync(region)
        .map(new Func1<RegionInfo, List<ModeInfo>>() {
          @Override public List<ModeInfo> call(RegionInfo regionInfo) {
            return regionInfo.transitModes();
          }
        });
  }
}

File: TripKitAndroid/src/main/java/com/skedgo/tripkit/RegionServiceImpl.kt
package com.skedgo.tripkit

import com.skedgo.tripkit.common.model.Location
import com.skedgo.tripkit.common.model.Region
import com.skedgo.tripkit.common.model.TransportMode
import com.skedgo.tripkit.data.regions.RegionService
import com.skedgo.tripkit.data.tsp.Paratransit
import com.skedgo.tripkit.data.tsp.RegionInfo
import com.skedgo.tripkit.tsp.RegionInfoRepository
import io.reactivex.Completable
import io.reactivex.Observable
import io.reactivex.Single
import com.skedgo.tripkit.routing.ModeInfo
import com.skedgo.tripkit.tsp.hasWheelChairInformation
import io.reactivex.functions.BiFunction

internal class RegionServiceImpl(
        private val regionCache: Cache<List<Region>>,
        private val modeCache: Cache<Map<String, TransportMode>>,
        private val regionsFetcher: RegionsFetcher,
        private val regionInfoRepository: RegionInfoRepository,
        private val regionFinder: RegionFinder) : RegionService {
    override fun getRegionsAsync(): Observable<List<Region>> = regionCache.async.toObservable()

    override fun getRegionByNameAsync(regionName: String): Observable<Region> =
            getRegionsAsync()
                    .flatMapIterable { it }
                    .filter { it.name == regionName }
                    .first(Region()).toObservable()

    override fun getTransportModesAsync(): Observable<Map<String, TransportMode>> =
            modeCache.async.toObservable()

    override fun getRegionByLocationAsync(latitude: Double, longitude: Double): Observable<Region> =
            getRegionsAsync()
                    .flatMap { regions ->
                        Observable.fromIterable(regions)
                    }
                    .filter { region ->
                        regionFinder.contains(region, latitude, longitude)
                    }
                    .firstOrError()
                    .onErrorResumeNext { error ->
                        if (error is NoSuchElementException)
                            Single.error<Region>(com.skedgo.tripkit.OutOfRegionsException(
                                    "Location lies outside covered area",
                                    latitude,
                                    longitude
                            ))
                        else
                            Single.error<Region>(error)
                    }.toObservable()

    override fun getRegionByLocationAsync(location: Location?): Observable<Region> =
            if (location == null)
                Observable.error<Region>(NullPointerException("Location is null"))
            else
                getRegionByLocationAsync(location.lat, location.lon)

    override fun getCitiesAsync(): Observable<Location> =
            getRegionsAsync()
                    .flatMap { regions -> Observable.fromIterable(regions) }
                    .compose(com.skedgo.tripkit.Utils.getCities())

    override fun getCitiesByNameAsync(name: String?): Observable<Location> =
            getCitiesAsync().filter(com.skedgo.tripkit.Utils.matchCityName(name))

    override fun getTransportModeByIdAsync(modeId: String): Observable<TransportMode> =
            getTransportModesAsync()
                    .flatMap {
                        val mode: TransportMode? = it[modeId]
                        mode?.let {
                            Observable.just(mode)
                        } ?: Observable.empty()
                    }

    override fun getTransportModesByIdsAsync(modeIds: List<String>): Observable<List<TransportMode>> =
            getTransportModesAsync().map(com.skedgo.tripkit.Utils.findModesByIds(modeIds))

    override fun getTransportModesByRegionAsync(region: Region): Observable<List<TransportMode>> {
        val modeIds = region.transportModeIds

        return if (modeIds != null)
            getTransportModesByIdsAsync(modeIds)
        else
            Observable.just(emptyList<TransportMode>())
    }


    override fun getTransportModesByLocationsAsync(location1: Location, location2: Location): Observable<List<TransportMode>> {
        return Observable.zip(getRegionByLocationAsync(location1), getRegionByLocationAsync(location2),
                BiFunction { first: Region, second: Region -> first to second })
                .flatMap { regionPair ->
                    Observable.zip(getRegionInfoByRegionAsync(regionPair.first), getRegionInfoByRegionAsync(regionPair.second),
                            BiFunction { regionInfoOne: RegionInfo, regionInfoTwo: RegionInfo ->
                                var transportModesOne = regionPair.first.transportModeIds
                                var transportModesTwo = regionPair.second.transportModeIds

                                if (regionInfoOne.hasWheelChairInformation() && transportModesOne?.contains(TransportMode.ID_WHEEL_CHAIR) != true) {
                                    transportModesOne?.add(TransportMode.ID_WHEEL_CHAIR)
                                }

                                if (regionInfoTwo.hasWheelChairInformation() && transportModesTwo?.contains(TransportMode.ID_WHEEL_CHAIR) != true) {
                                    transportModesTwo?.add(TransportMode.ID_WHEEL_CHAIR)
                                }
                                transportModesOne to transportModesTwo
                            })
                }
                .flatMap {
                    val list1 = it.first ?: arrayListOf<String>()
                    val list2 = it.second ?: arrayListOf<String>()

                    getTransportModesByIdsAsync(list1.union(list2).toList())
                }
    }


    override fun getTransportModesByLocationAsync(
            location: Location
    ): Observable<List<TransportMode>> =
            getRegionByLocationAsync(location)
                    .flatMap { region ->
                        val modeIds = region.transportModeIds
                        if (modeIds != null)
                            getTransportModesByIdsAsync(modeIds)
                        else
                            Observable.just(emptyList<TransportMode>())
                    }

    override fun refreshAsync(): Completable =
            regionsFetcher.fetchAsync()
                    .doOnComplete {
                        regionCache.invalidate()
                        modeCache.invalidate()
                        regionFinder.invalidate()
                    }

    override fun fetchParatransitByRegionAsync(region: Region): Observable<Paratransit> =
            getRegionInfoByRegionAsync(region)
                    .map { regionInfo -> regionInfo.paratransit() }

    override fun getRegionInfoByRegionAsync(region: Region): Observable<RegionInfo> =
            regionInfoRepository.getRegionInfoByRegion(region)

    override fun getTransitModesByRegionAsync(region: Region): Observable<List<ModeInfo>> =
            getRegionInfoByRegionAsync(region)
                    .map { regionInfo -> regionInfo.transitModes() }
}
